nuagekubemon/client/nuageetcdclient.go

```diff
@@ -0,0 +1,624 @@
+			if err != nil {
+				glog.Errorf("getting response from etcd failed: %v", err)
+				return err
+			}
+			return nil
+		})
+	if err != nil {
+		return subnetList, err
+	}
+	if len(getResp.Kvs) != 1 {
+		glog.Errorf("received more than one kv pairs")
+		return subnetList, err
+	}
+	defer nuageEtcdRetry(
+		func() error {
+			_, err = nuageetcd.client.Delete(context.Background(), POD_METADATA_TREE+ns+"/"+pod)
+			if err != nil {
+				glog.Errorf("deleting pod metadata from etcd failed: %v", err)
+				return err
+			}
+			return nil
+		})
+
+	subnetStr := string(getResp.Kvs[0].Value)
+	f := func(s conc.STM) error {
+		snet := &etcdSubnetValue{}
+		if err := json.Unmarshal([]byte(s.Get(SUBNET_METADATA_TREE+ns+"/"+subnetStr)), snet); err != nil {
+			glog.Errorf("unmarshal string %s to struct failed: %v", s.Get(SUBNET_METADATA_TREE+ns+"/"+subnetStr), err)
+			return err
+		}
+		snet.ACTIVEIP -= 1
+		b, err := json.Marshal(snet)
+		if err != nil {
+			glog.Errorf("marshal struct %v to string failed: %v", snet, err)
+			return err
+		}
+		s.Put(SUBNET_METADATA_TREE+ns+"/"+subnetStr, string(b))
+
+		ACTIVEIPCount := 0
+		noOfSubnets := 0
+		for _, key := range s.GetKeys() {
+			if err := json.Unmarshal([]byte(s.Get(key)), snet); err != nil {
+				glog.Errorf("unmarshal json string(%s) to struct failed: %v", s.Get(key), err)
+			} else {
+				ACTIVEIPCount += snet.ACTIVEIP
+				noOfSubnets += 1
+			}
+		}
+
+		if noOfSubnets != 1 && ACTIVEIPCount*100 < noOfSubnets*nuageetcd.maxIPCount*SCALE_DOWN_THRESHOLD {
+			for _, key := range s.GetKeys() {
+				if err := json.Unmarshal([]byte(s.Get(key)), snet); err != nil {
+					glog.Errorf("unmarshal json string(%s) to struct failed: %v", s.Get(key), err)
+				} else {
+					if snet.ACTIVEIP == 0 {
+						s.Del(key)
+						subnetList = append(subnetList, &api.EtcdSubnetMetadata{ID: snet.VSDID, CIDR: snet.CIDR})
+					}
+				}
+			}
+		}
+
+		return nil
+	}
+	if err := nuageetcd.nuageSTM(f, ns+"/"+ns); err != nil {
+		glog.Errorf("De-allocating pod ip from subnet failed with error: %v", err)
+		return subnetList, err
+	}
+	return subnetList, nil
+}
+
+//CreateFirstSubnet creates the first subnet in a namespace
+func (nuageetcd *NuageEtcdClient) CreateFirstSubnet(subnetInfo *api.EtcdSubnetMetadata) error {
+	var err error
+	var getResp *clientv3.GetResponse
+	b := []byte{}
+	ns := subnetInfo.Namespace
+	subnet := subnetInfo.Name
+	if err = nuageetcd.nuageMutex(func() error {
+		nuageEtcdRetry(
+			func() error {
+				getResp, err = nuageetcd.client.Get(context.Background(), SUBNET_METADATA_TREE+ns+"/"+subnet)
+				if err != nil {
+					glog.Errorf("get on %s failed: %v", SUBNET_METADATA_TREE+ns+"/"+subnet, err)
+					return err
+				}
+				return nil
+			})
+		if err != nil {
+			return err
+		}
+		if len(getResp.Kvs) != 0 {
+			return nil
+		}
+		s := &etcdSubnetValue{ACTIVEIP: 0, VSDID: subnetInfo.ID, CIDR: subnetInfo.CIDR}
```

Comment:
yes. for creating a first subnet in a namespace VSD operations are done first. This method takes a subnetInfo structure which has the VSD ID once it is created. so the order of operations is this
1) only one master creates subnet on VSD
2) master will create subnet in etcd if it is not already present.